cpp/src/sampling/detail/prepare_next_frontier_impl.cuh

```diff
@@ -0,0 +1,195 @@
+    if (frontier_vertex_labels) {
+      std::tie(frontier_vertices, *frontier_vertex_labels) =
+        shuffle_int_vertex_value_pairs_to_local_gpu_by_vertex_partitioning(
+          handle,
+          std::move(frontier_vertices),
+          std::move(*frontier_vertex_labels),
+          vertex_partition_range_lasts);
+    } else {
+      frontier_vertices = shuffle_int_vertices_to_local_gpu_by_vertex_partitioning(
+        handle, std::move(frontier_vertices), vertex_partition_range_lasts);
+    }
+  }
+
+  if (frontier_vertex_labels) {
+    auto begin_iter =
+      thrust::make_zip_iterator(frontier_vertices.begin(), frontier_vertex_labels->begin());
+    thrust::sort(handle.get_thrust_policy(), begin_iter, begin_iter + frontier_vertices.size());
+  } else {
+    thrust::sort(handle.get_thrust_policy(), frontier_vertices.begin(), frontier_vertices.end());
+  }
+
+  if (vertex_used_as_source) {
+    auto& [verts, labels] = *vertex_used_as_source;
+
+    // add sources from this expansion to the vertex_used_as_source
+    size_t current_verts_size = verts.size();
+    size_t new_verts_size     = current_verts_size + sampled_src_vertices.size();
+
+    verts.resize(new_verts_size, handle.get_stream());
+
+    raft::copy(verts.begin() + current_verts_size,
+               sampled_src_vertices.data(),
+               sampled_src_vertices.size(),
+               handle.get_stream());
+
+    // sort and unique the vertex_used_as_source structures
+    if (sampled_src_vertex_labels) {
+      labels->resize(new_verts_size, handle.get_stream());
+
+      raft::copy(labels->begin() + current_verts_size,
+                 sampled_src_vertex_labels->data(),
+                 sampled_src_vertex_labels->size(),
+                 handle.get_stream());
+
+      auto begin_iter = thrust::make_zip_iterator(verts.begin(), labels->begin());
+
+      thrust::sort(handle.get_thrust_policy(), begin_iter, begin_iter + new_verts_size);
+
+      auto end_iter =
+        thrust::unique(handle.get_thrust_policy(), begin_iter, begin_iter + new_verts_size);
+
+      verts.resize(thrust::distance(begin_iter, end_iter), handle.get_stream());
+      labels->resize(thrust::distance(begin_iter, end_iter), handle.get_stream());
+    } else {
+      thrust::sort(handle.get_thrust_policy(), verts.begin(), verts.end());
+
+      auto end_iter = thrust::unique(handle.get_thrust_policy(), verts.begin(), verts.end());
+
+      verts.resize(thrust::distance(verts.begin(), end_iter), handle.get_stream());
+    }
+
+    // Now with the updated verts/labels we can filter the next frontier
+    std::tie(frontier_vertices, frontier_vertex_labels) = remove_visited_vertices_from_frontier(
+      handle,
+      std::move(frontier_vertices),
+      std::move(frontier_vertex_labels),
+      raft::device_span<vertex_t const>{verts.data(), verts.size()},
+      labels ? std::make_optional(raft::device_span<label_t const>{labels->data(), labels->size()})
+             : std::nullopt);
```

Comment:
Shouldn't we do this only when `EXCLUDE` is selected?

Reply:
As mentioned above, `EXCLUDE` being selected and `vertex_used_as_source` being defined are synonymous.